crates/core/src/repository/command_input.rs

```diff
@@ -0,0 +1,215 @@
+    //    deserialize_with = "deserialize_command"
+    //)]
+    #[serde_as(as = "PickFirst<(DisplayFromStr,_)>")] CommandInputInternal,
+);
+
+impl Serialize for CommandInput {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        if self.0.on_failure.is_none() || self.0.on_failure == Some(OnFailure::default()) {
+            serializer.serialize_str(&self.to_string())
+        } else {
+            self.0.serialize(serializer)
+        }
+    }
+}
+
+impl From<Vec<String>> for CommandInput {
+    fn from(value: Vec<String>) -> Self {
+        Self(CommandInputInternal::from_vec(value))
+    }
+}
+
+impl From<CommandInput> for Vec<String> {
+    fn from(value: CommandInput) -> Self {
+        value.0.iter().cloned().collect()
+    }
+}
+
+impl CommandInput {
+    /// Returns if a command is set
+    #[must_use]
+    pub fn is_set(&self) -> bool {
+        self.0.command.is_some()
+    }
+
+    /// Returns the command if it is set
+    ///
+    /// # Panics
+    ///
+    /// Panics if no command is set.
+    #[must_use]
+    pub fn command(&self) -> &str {
+        self.0.command.as_ref().unwrap()
+    }
+
+    /// Returns the command args if it is set
+    #[must_use]
+    pub fn args(&self) -> &[String] {
+        &self.0.args.0
+    }
+
+    /// Runs the command if it is set
+    ///
+    /// # Errors
+    ///
+    /// `RusticError` if return status cannot be read
+    pub fn run(&self, context: &str, what: &str) -> RusticResult<()> {
+        if !self.is_set() {
+            trace!("not calling command {context}:{what} - not set");
+            return Ok(());
+        }
+        debug!("calling command {context}:{what}: {self:?}");
+        let status = Command::new(self.command())
+            .args(self.args())
+            .status()
+            .map_err(|err| {
+                RepositoryErrorKind::CommandExecutionFailed(context.into(), what.into(), err).into()
+            });
+        let Some(status) = eval_on_failure(&self.0.on_failure, status)? else {
+            return Ok(());
+        };
+
+        if !status.success() {
+            let _: Option<()> = eval_on_failure(
+                &self.0.on_failure,
+                Err(
+                    RepositoryErrorKind::CommandErrorStatus(context.into(), what.into(), status)
+                        .into(),
+                ),
+            )?;
+        }
+        Ok(())
+    }
+}
+
+impl FromStr for CommandInput {
+    type Err = RusticError;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Ok(Self(CommandInputInternal::from_str(s)?))
+    }
+}
+
+impl Display for CommandInput {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        Display::fmt(&self.0, f)
+    }
+}
+
+#[serde_as]
+#[derive(Debug, Default, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(default, rename_all = "kebab-case")]
+struct CommandInputInternal {
+    command: Option<String>,
```

Comment:
`command` is now just `String`. In case users do give an empty command to parse, the `command` is an empty string which we can check on.

But I agree that using `Option<CommandInput>` or `Vec<CommandInput>` is what should be used for optional commands.